### PR TITLE
NEXT-38822 - Filter list and checkout endpoints

### PR DESCRIPTION
**Pull request by Bird87ZA**
Added the gateway between listing and checkout.

Gateway filters invalid options and only allows checkout and listing of purchases that can actually be purchased.

### CODE REVIEW

File: src/Services/InAppPurchasesService.php [hunk 1]
```diff
@@ -33,4 +36,31 @@ public function listPurchases(string $extensionName, Context $context): InAppPur
...
+            return $allPurchases;
+        }
+
+        foreach ($allPurchases as $key => $purchase) {
+            if (\in_array($purchase->getIdentifier(), $validPurchases)) {
+                continue;
+            }
+
+            $allPurchases->remove($key);
+        }
```
Comment (cyl3x) [inline src/Services/InAppPurchasesService.php]: The `Collection` struct has a filter method

File: src/Struct/InAppPurchaseCartPositionCollection.php [hunk 2]
```diff
@@ -41,4 +41,12 @@ protected function getExpectedClass(): ?string
     {
         return InAppPurchaseCartPositionStruct::class;
     }
+
+    /**
+     * @return array<int, string>
+     */
+    public function toArray(): array
+    {
+        return array_map(static fn(InAppPurchaseCartPositionStruct $element) => $element->getInAppFeatureIdentifier(), $this->elements);
```
Comment (cyl3x) [inline src/Struct/InAppPurchaseCartPositionCollection.php]: The `Collection` struct has a map method

File: src/Struct/InAppPurchaseCollection.php [hunk 3]
```diff
@@ -26,4 +26,12 @@ public static function fromArray(array $data): self
 
         return new self($elements);
     }
+
+    /**
+     * @return array<int, string>
+     */
+    public function toArray(): array
+    {
+        return array_map(static fn(InAppPurchaseStruct $element) => $element->getIdentifier(), $this->elements);
```
Comment (cyl3x) [inline src/Struct/InAppPurchaseCollection.php]: s.a.

File: src/Services/InAppPurchasesService.php [hunk 4]
```diff
@@ -33,4 +36,31 @@ public function listPurchases(string $extensionName, Context $context): InAppPur
     {
         return $this->client->listInAppPurchases($extensionName, $context);
     }
+
+    /**
+     * @param Collection<InAppPurchaseCartPositionStruct>|Collection<InAppPurchaseStruct> $allPurchases
```
Comment (cyl3x) [inline src/Services/InAppPurchasesService.php]: Doesn't it make sense to use your Interface here?
```suggestion
     * @param Collection<InAppPurchaseIdentifiableStruct> $allPurchases
```

**Review by cyl3x — CHANGES_REQUESTED** [review]

File: src/Services/InAppPurchasesService.php (hunk 4 above)
Comment (Bird87ZA) [reply src/Services/InAppPurchasesService.php]: I had it like that, but then phpstan complained. Changed it again, and saw it's easily solvable with `covariants`

File: src/Struct/InAppPurchaseCollection.php [hunk 5]
```diff
@@ -26,4 +26,12 @@ public static function fromArray(array $data): self
 
         return new self($elements);
     }
+
+    /**
+     * @return array<int, string>
+     */
+    public function toArray(): array
```
Comment (cyl3x) [inline src/Struct/InAppPurchaseCollection.php]: I think it's misleading to call this `toArray' when it actually returns a list of identifiers.

File: src/Struct/InAppPurchaseCartPositionCollection.php [hunk 6]
```diff
@@ -41,4 +41,12 @@ protected function getExpectedClass(): ?string
     {
         return InAppPurchaseCartPositionStruct::class;
     }
+
+    /**
+     * @return array<int, string>
+     */
+    public function toArray(): array
```
Comment (cyl3x) [inline src/Struct/InAppPurchaseCartPositionCollection.php]: s.a.

**Review by cyl3x — CHANGES_REQUESTED** [review]

**Review by cyl3x — APPROVED** [review]

File: src/Controller/InAppPurchasesController.php [hunk 7]
```diff
@@ -62,6 +66,15 @@ public function orderCart(RequestDataBag $data, Context $context): Response
 
         $positionCollection = InAppPurchaseCartPositionCollection::fromArray($positions->all());
 
+        $validCartItems = $this->appPurchasesGateway->process(new InAppPurchasesPayload($positionCollection->getIdentifiers()));
+
+        $positionCollection = $this->inAppPurchasesService->filterValidInAppPurchases($positionCollection, $validCartItems->getPurchases());
+        \assert($positionCollection instanceof InAppPurchaseCartPositionCollection);
```
Comment (mstegmeyer) [inline src/Controller/InAppPurchasesController.php]: with the other comments of mine, you hopefully don't need this assert

File: src/Services/InAppPurchasesService.php [hunk 8]
```diff
@@ -33,4 +37,27 @@ public function listPurchases(string $extensionName, Context $context): InAppPur
...
+    /**
+     * @param Collection<covariant InAppPurchaseIdentifiableStruct> $allPurchases
+     * @param array<int, string> $validPurchases
+     *
+     * @return Collection<covariant InAppPurchaseIdentifiableStruct>
+     */
+    public function filterValidInAppPurchases(
+        Collection $allPurchases,
+        array $validPurchases,
+    ): Collection {
```
Comment (mstegmeyer) [inline src/Services/InAppPurchasesService.php]: I find this function way to complicated with the covariant stuff (and the necessary asserts). I would just do this as separate functions of each of the two collections. Then you also don't need the interface.

File: src/Services/InAppPurchasesService.php [hunk 9]
```diff
@@ -33,4 +37,27 @@ public function listPurchases(string $extensionName, Context $context): InAppPur
...
+        if (empty($validPurchases)) {
+            $allPurchases->clear();
+
+            return $allPurchases;
+        }
+
+        $allPurchases->filter(function (InAppPurchaseStruct|InAppPurchaseCartPositionStruct $purchase) use ($validPurchases) {
```
Comment (mstegmeyer) [inline src/Services/InAppPurchasesService.php]: I can't really imagine that this works. `filter` in Shopware Collections returns a new Collection and does not manipulate the old one. Therefore, in your empty case above, you manipulate the existing collection and return it, and here, you create a new one, ignore it and return the unmanipulated one.

**Review by mstegmeyer — CHANGES_REQUESTED** [review]

File: src/Struct/InAppPurchaseCartPositionStruct.php [hunk 10]
```diff
@@ -31,7 +31,9 @@ private function __construct(
     public static function fromArray(array $data): self
     {
         $inAppPurchaseCartPosition = (new self())->assign($data);
-        $inAppPurchaseCartPosition->setInAppFeatureIdentifier($data['feature']['identifier'] ?? '');
+        if ($inAppPurchaseCartPosition->getInAppFeatureIdentifier() === '') {
```
Comment (cyl3x) [inline src/Struct/InAppPurchaseCartPositionStruct.php:34]: Why is this necessary? It seems unexpected that `fromArray' does not allow you to override the identifier

**Review by mstegmeyer — APPROVED** [review]

File: src/Struct/InAppPurchaseCartPositionStruct.php (hunk 10 above)
Comment (Bird87ZA) [reply src/Struct/InAppPurchaseCartPositionStruct.php:34]: If the data received already has `in-app-feature-identifier` then it will overwrite that value with empty if `$data['feature']['identifier']` is not set.

This is a fallback for when `in-app-feature-identifier` doesn't exist in the array.

Comment (cyl3x) [reply src/Struct/InAppPurchaseCartPositionStruct.php:34]: doesn't it make sense then to check on the data instead of the object? `if ($data['feature']['identifier']) { ... }` or `if (!$data['in-app-feature-identifier']) { ... }`

Comment (Bird87ZA) [reply src/Struct/InAppPurchaseCartPositionStruct.php:34]: Well no, I only want to fall back it if the object doesn't have `$inAppFeatureIdentifier` set. If that is set, I don't want to overwrite it with `$data['feature']['identifier']`.

**Review by cyl3x — APPROVED** [review]

**Review by mstegmeyer — APPROVED** [review]
